ass/document.py

```diff
@@ -578,3 +432,247 @@ class Command(_Event):
+
+    def dump(self):
+        yield "[{}]".format(self.section)
+
+        for k, v in self._fields.items():
+            yield "{}: {}".format(k, _Field.dump(v))
+
+    def __getitem__(self, key):
+        return self._fields[key]
+
+    def __setitem__(self, key, value):
+        self._fields[key] = value
+
+    def __delitem__(self, key):
+        del self._fields[key]
+
+    def __iter__(self):
+        return iter(self._fields)
+
+    def __len__(self):
+        return len(self._fields)
+
+
+class EventsSection(LineSection):
+    FIELD_ORDER = Dialogue.DEFAULT_FIELD_ORDER
+    LINE_PARSERS = {
+        "dialogue": Dialogue,  # noqa: E241
+        "comment":  Comment,   # noqa: E241
+        "picture":  Picture,   # noqa: E241
+        "sound":    Sound,     # noqa: E241
+        "movie":    Movie,     # noqa: E241
+        "command":  Command    # noqa: E241
+    }
+
+class StylesSection(LineSection):
+    FIELD_ORDER = Style.DEFAULT_FIELD_ORDER
+    LINE_PARSERS = {
+        "style": Style
+    }
+
+class ScriptInfoSection(FieldSection):
+    VERSION_ASS = "v4.00+"
+    VERSION_SSA = "v4.00"
+    FIELDS = {
+        "ScriptType": _Field("ScriptType", str, default=VERSION_ASS),
+        "PlayResX": _Field("PlayResX", int, default=640),
+        "PlayResY": _Field("PlayResY", int, default=480),
+        "WrapStyle": _Field("WrapStyle", int, default=0),
+        "ScaledBorderAndShadow": _Field("ScaledBorderAndShadow", str, default="yes")
+    }
+
+
+@add_metaclass(_WithFieldMeta)
+class Document(object):
+    """ An ASS document. """
+    SCRIPT_INFO_HEADER = "Script Info"
+    STYLE_SSA_HEADER = "V4 Styles"
+    STYLE_ASS_HEADER = "V4+ Styles"
+    EVENTS_HEADER = "Events"
+    AEGISUB_PROJECT_HEADER = "Aegisub Project Garbage"
+
+    SECTIONS = CaseInsensitiveOrderedDict({
+        SCRIPT_INFO_HEADER: ScriptInfoSection,
+        AEGISUB_PROJECT_HEADER: FieldSection,
+        STYLE_SSA_HEADER: StylesSection,
+        STYLE_ASS_HEADER: StylesSection,
+        EVENTS_HEADER: EventsSection
+    })
+
+    # backwards compatibility
+    script_type = ScriptInfoSection.FIELDS["ScriptType"]
+    play_res_x = ScriptInfoSection.FIELDS["PlayResX"]
+    play_res_y = ScriptInfoSection.FIELDS["PlayResY"]
+    wrap_style = ScriptInfoSection.FIELDS["WrapStyle"]
+    scaled_border_and_shadow = ScriptInfoSection.FIELDS["ScaledBorderAndShadow"]
+
+    def __init__(self):
+        """ Create an empty ASS document.
+        """
+        self.sections = CaseInsensitiveOrderedDict(
+            [(header, self.SECTIONS[header](header)) for header in (self.SCRIPT_INFO_HEADER,
+                                                                    self.STYLE_ASS_HEADER,
+                                                                    self.EVENTS_HEADER)])
+        self.fields = self.sections[self.SCRIPT_INFO_HEADER]
+        self.styles = self.sections[self.STYLE_ASS_HEADER]
+        self.events = self.sections[self.EVENTS_HEADER]
```

Comment:
The way this is implemented is a little error-prone because they aren't transparent to assignments to these attributes, `sections` or `sections`'s items. E.g. in `parse_file` you need to specifically check for existance of these.

I would implement them as properties instead.

Reply:
You're right, it doesn't handle assignment at the moment. And it turns out that I'm actually relying on that in my very own Sub Digest. I guess a property that can explicitly convert a list/dict to the appropriate section type on assignment is the way to go.